fJVT scale factors support

ntadej opened the pull request:
This PR fixes fJVT when using systematics and adds fJVT scale factors.

It might be good idea for the future to move fJVT to `JetCalibrator` to avoid the need of deep copying.

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -236,13 +236,47 @@ EL::StatusCode JetSelector :: initialize ()
+  if (m_dofJVT) {
+    ANA_MSG_DEBUG("Trying to initialize fJVT tool");
+    //ANA_CHECK( ASG_MAKE_ANA_TOOL(m_fJVT_tool_handle, JetForwardJvtTool));
+    ANA_CHECK(m_fJVT_tool_handle.retrieve());
```
missing a setToolName call here (we need it to make sure the tool is effective private)

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -236,13 +236,47 @@ EL::StatusCode JetSelector :: initialize ()
+    ANA_MSG_DEBUG("Successfully initialized fJVT tool");
+
+    // initialize the CP::JetJvtEfficiency Tool for fJVT
+    setToolName(m_fJVT_eff_tool_handle, "fJVT_eff_tool");
```
Don't need to set the tool name here (no second argument) unless there's more than one tool in this algorithm for it, or if you need to share it with another algorithm

ntadej replied:
There are two instances of tools, one for JVT and one for fJVT. If I do not set the names, the same one is used both times (the one which is initialised first).

kratsg replied:
Good point. Wasn't obvious at first but now I see it.

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -236,13 +236,47 @@ EL::StatusCode JetSelector :: initialize ()
-  // initialize the CP::JetJvtEfficiency Tool
-  setToolName(m_JVT_tool_handle);
+  // initialize the CP::JetJvtEfficiency Tool for JVT
+  setToolName(m_JVT_tool_handle, "JVT_eff_tool");
```
don't need to set name unless you need to share with another algorithm.

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
+          sfVecfJVT( *jet ).push_back( fjvtSF );
+
+          ANA_MSG_DEBUG( "===>>>");
+          ANA_MSG_DEBUG( " ");
```
no need for empty line

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
+          ANA_MSG_DEBUG( "===>>>");
+          ANA_MSG_DEBUG( " ");
+          ANA_MSG_DEBUG( "Jet " << idx << ", pt = " << jet->pt()*1e-3 << " GeV, |eta| = " << std::fabs(jet->eta()) );
+          ANA_MSG_DEBUG( " ");
```
empty line

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
+          ANA_MSG_DEBUG( "Jet " << idx << ", pt = " << jet->pt()*1e-3 << " GeV, |eta| = " << std::fabs(jet->eta()) );
+          ANA_MSG_DEBUG( " ");
+          ANA_MSG_DEBUG( "fJVT SF decoration: " << m_outputSystNamesfJVT );
+          ANA_MSG_DEBUG( " ");
```
empty line

kratsg commented on Root/JetSelector.cxx:
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
+          ANA_MSG_DEBUG( "fJVT SF decoration: " << m_outputSystNamesfJVT );
+          ANA_MSG_DEBUG( " ");
+          ANA_MSG_DEBUG( "Systematic: " << syst_it.name() );
+          ANA_MSG_DEBUG( " ");
```
empty line

kratsg commented on xAODAnaHelpers/JetSelector.h:
```diff
@@ -165,6 +165,38 @@ class JetSelector : public xAH::Algorithm
+        "Medium"  87.1-97.0%     53.4-60.9%
+        "Tight"   79.9-95.6%     45.4-50.3%
+        ======== ============== =============
+
```
can you add a link to where the above numbers are from?

kratsg commented on xAODAnaHelpers/JetSelector.h:
```diff
@@ -165,6 +165,38 @@ class JetSelector : public xAH::Algorithm
+     @brief Configuration containting fJVT scale factors.
+
+     @rst
+     The configuration file with the scale factors calculated by the ``CP::IJetJvtEfficiency``.
```
the text inside `@rst/..../@endrst` needs to be indented at least once.

ntadej commented:
I've also noticed that we need passed info for data so I added it. I'm open for suggestions where to put accessors in JetContainer now.

kratsg commented:
> I've also noticed that we need passed info for data so I added it. I'm open for suggestions where to put accessors in JetContainer now.

This seems fine for right now. We might want to provide a uniform area for all accessors in our code, but I suspect this is a long term goal / cleanup effort.

ntadej commented:
I addressed all of your comments. I can rebase if needed.

The only thing that is still on my mind is to check if deep copies are really necessary for fJVT, or should I open a new MR for it?

ntadej commented:
I updated branches to be of type `std::vector<int>` not `std::vector<char>` + rebased.

Also it seems that fJVT in current place needs deep copies so I will try to fix that in a separate PR.

ntadej commented:
Hi @kratsg, is there anything else that needs to be done from my side for this to be merged?

kratsg commented:
> Hi @kratsg, is there anything else that needs to be done from my side for this to be merged?

Nope, but thanks for poking me. I've merged it in :)

kratsg commented:
I also approved the changes (but forgot to mark that on the review, sorry).
